### ISSUE TITLE
[no-conflicting-lifecycle] false positives

### ISSUE THREAD

**Issue by rafaelss95** [angular-eslint/angular-eslint#502]
**Description and reproduction of the issue**
As mentioned in https://github.com/mgechev/codelyzer/issues/560#issuecomment-621313192 and https://github.com/mgechev/codelyzer/issues/588:

> no-conflicting-life-cycle-hooks should prevents this:

> Note that a directive typically should not use both DoCheck and OnChanges to respond to changes **on the same input**, as ngOnChanges will continue to be called when the default change detector detects changes.

> Currently, it doesn't seem to be the case. We just check if we use the both hooks in same time. 

This rule should report **only** if a variable is being checked in both `ngDoCheck` and` ngOnChanges` methods, not as it currently does that reports if you implement both `DoCheck` and `OnChanges` or have corresponding methods.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@angular-eslint       `        | `12.0.0` |
| `TypeScript`                       | `4.3.0` |
| `ESLint`                           | `7.27.0` |

```sh
# Please run `npx ng version` in your project and paste the full output here:
Angular CLI: 12.0.0
Node: 14.16.0
Package Manager: npm 6.14.11
OS: darwin x64

Angular: 12.0.0
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, language-service, localize, material
... platform-browser, platform-browser-dynamic, platform-server
... router, service-worker

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1200.0
@angular-devkit/build-angular   12.0.0
@angular-devkit/core            11.2.11
@angular-devkit/schematics      11.2.11
@nguniversal/builders           12.0.0
@nguniversal/common             12.0.0
@nguniversal/express-engine     12.0.0
@schematics/angular             12.0.0
ng-packagr                      12.0.0
rxjs                            6.5.5
typescript                      4.2.4

```

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest supported version of the packages and checked my `ng version` output per the instructions given here.

**Comment by rafaelss95** [comment]
Being quite honest: this is a very difficult (not to say impossible) rule to implement correctly, due to the many possibilities.

What actually defines whether an input is being checked in both methods at all? Should the code below trigger the rule? If so, why?

```ts
@Component()
class ChildComponent implements DoCheck {
  @Input() person!: Person;
  readonly differ = this.keyValueDiffers.find({}).create();
  lastId?: string;

  constructor(private readonly keyValueDiffers: KeyValueDiffers) {}

  ngOnChanges() {
    this.lastId = this.person.id;
  }

  ngDoCheck() {
    const changes = this.differ.diff(this.person);

    if (!changes) {
      return;
    }

    changes.forEachChangedItem(({ currentValue }) =>
      console.log('changed: ', currentValue),
    );
    changes.forEachAddedItem(({ currentValue }) =>
      console.log('added: ', currentValue),
    );
    changes.forEachRemovedItem(({ currentValue }) =>
      console.log('removed: ', currentValue),
    );
  }
}

```

**Comment by JamesHenry** [comment]
Apologies folks, because of personal life pressures I have had to limit angular-eslint to v13 of Angular and v8 of ESLint, so this breaking change requirement will be bumped to v14

**Comment by devoto13** [comment]
IMO this rule does not make any sense and should be removed together with the confusing statement in the [Angular docs](https://angular.io/api/core/DoCheck#usage-notes):

> When the default change detector detects changes, it invokes ngOnChanges() if supplied, regardless of whether you perform additional change detection. Typically, you should not use both [DoCheck](https://angular.io/api/core/DoCheck) and [OnChanges](https://angular.io/api/core/OnChanges) to respond to changes on the same input.

Typically, nobody uses `ngDoCheck` to respond to input changes as there `ngOnChanges` for it. Typically, one should not use `ngDoCheck` at all unless they know exactly what they are doing.

**Comment by JamesHenry** [comment]
The initial decision, and breaking change, I am going to make on this one, is that in v15 this rule will no longer be part of the recommended config exported by `@angular-eslint/eslint-plugin`.

This is in light of the various comments above concluding that it is not universally applicable and not necessarily doing a good job of fulfilling its intended purpose.

As @rafaelss95 points out, its intended purpose may well be very difficult to actually implement consistently, and we may have to consider full deprecation of this rule in future.

**Comment by JamesHenry** [comment]
I'm going to formally deprecate this rule in the upcoming v21, and remove it entirely in v22

**Comment by JamesHenry** [comment]
It's deprecated and removed from the "all" config in https://www.npmjs.com/package/angular-eslint/v/21.0.0

**Comment by JamesHenry** [comment]
Removed in v22
